### PR TITLE
Add metrics framework and network time metric

### PR DESCRIPTION
**Pull request by maxjeffos**
- [X] Ready for review
- [X] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules
- [x] Reviewed by Snyk internal team

#### What does this PR do?
- adds a metrics collecting framework so that we can time things, etc
- collects total network time
- collects "cpu time" which is roughly estimated as total time - network time, understanding that it is not quite that simple

#### Any background context you want to provide?
[Corresponding PR on registry](https://github.com/snyk/registry/pull/16616)

#### What are the relevant tickets?
- HAMMER-189
- the motivation was to have some data for when we go and tackle other improvements/fixes, including HAMMER-189

### CODE REVIEW

File: src/lib/request/index.ts [hunk 1]
```diff
@@ -21,5 +27,8 @@ export = async (
       return callback(error);
     }
     throw error;
+  } finally {
+    // call stop again here in case there's an error and the first .stop() is skipped over
+    totalNetworkTimeTimer.stop();
```
Comment (JackuB) [inline src/lib/request/index.ts:32]: Shouldn't a single `stop` call in `finally` be enough? You could say that even the response processing is part of the network

File: src/lib/analytics.js [hunk 2]
```diff
@@ -79,6 +80,7 @@ function postAnalytics(data) {
 
       data.ci = isCI();
       data.durationMs = Date.now() - startTime;
+      data.metrics = MetricsCollector.getInstance().getAllMetricsData();
```
Comment (JackuB) [inline src/lib/analytics.js:83]: I think I'd prefer we set these keys explicitly and prevent (accidental) dynamic fields loaded

File: src/lib/metrics.ts [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+  }
+}
+
+export class NamedMetrics {
+  public static NETWORK_TIME: MetricId = new MetricId(
+    'network-time',
+    'amount of time in ms spent waiting on network requests',
+    METRIC_TYPE_TIMER,
+  );
+}
```
Comment (JackuB) [inline src/lib/metrics.ts:208]: I'm a bit worried that this file is 200 lines alone 🤔 is there something to simplify the interface? Maybe moving the Counter from standalone class?

File: src/lib/metrics.ts [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    if (this.startTimeMs === 0) {
+      this.startTimeMs = Date.now();
+    } else {
+      MetricsCollector.getInstance().logImproperUse(
+        'Invalid Timer use: start() called when timer already stopped',
+      );
+    }
+  }
+
+  public stop() {
```
Comment (JackuB) [inline src/lib/metrics.ts:175]: I'd consider debug-logging the timer and its value as well

File: src/lib/request/index.ts (hunk 1 above)
Comment (maxjeffos) [reply src/lib/request/index.ts:32]: I was just trying to get the .stop() as close to the network bits as possible and didn't to take those callbacks at line 22 into the timer time. But, sure you could consider processing them as part of the timer - would make it so we just call the timer in the finally which looks prettier.

File: src/lib/metrics.ts (hunk 3 above)
Comment (maxjeffos) [reply src/lib/metrics.ts:208]: I'm not against breaking up the this metrics into multiple files. I think that would be best.

File: src/lib/analytics.js (hunk 2 above)
Comment (maxjeffos) [reply src/lib/analytics.js:83]: It's very explicit - all keys need to be hard coded in the NameMetrics class.

File: src/lib/metrics.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    if (this.endTimeMs === 0) {
+      this.endTimeMs = Date.now();
+      debug(
+        `Timer stopped at ${
```
Comment (JackuB) [inline src/lib/metrics.ts:38]: This wouldn't tell you which timer, no?

File: src/lib/metrics.ts [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        name: m.name,
+        type: m.metricType,
+        values: m.getValues(),
+        total: m.getTotal(),
+      };
+      res.push(nextEntry);
+    });
+    return res;
+  }
+}
```
Comment (JackuB) [inline src/lib/metrics.ts:142]: You could probably map these:

```suggestion
  public static getAllMetrics(): any[] {
    const metrics: Metric[] = [
      MetricsCollector.NETWORK_TIME,
      MetricsCollector.CPU_TIME,
    ];
    return metrics.map(m => ({
      name: m.name,
      type: m.metricType,
      values: m.getValues(),
      total: m.getTotal(),
    }));
  }
}
```

File: src/lib/metrics.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    return sm;
+  }
+}
+
+export function logImproperUse(msg: string) {
+  debug(msg);
+}
+
+export class MetricsCollector {
+  public static NETWORK_TIME: TimerMetric = new TimerMetric(
```
Comment (JackuB) [inline src/lib/metrics.ts:114]: Never did too much OOP - what's the reason to have both `TimerMetric` and `TimerMetricInstance`?

File: src/lib/metrics.ts (hunk 6 above)
Comment (maxjeffos) [reply src/lib/metrics.ts:142]: Thanks

File: src/lib/metrics.ts (hunk 7 above)
Comment (maxjeffos) [reply src/lib/metrics.ts:114]: `TimerMetricInstance` is the individual thing where the `.start()` and `.stop()` and elapsed time calculation happen (in the `.getValue()`). `TimerMetric` is an abstraction that allows you to have _n_ `TimerMetricInstances`, thus allowing us to have multiple runs of the "same" metric (as we do for `NETWORK-TIME`), be able to get the total, and collect each individual value.

File: src/lib/metrics.ts (hunk 5 above)
Comment (maxjeffos) [reply src/lib/metrics.ts:38]: Yes. Which is not ideal. I didn't want to kludge up (de-normalize, more specifically) the thing by passing the context and type right down into each individual TimerMetricInstance. But I think it's worth it to have it in the log and I thought of a less ugly way to do it. Will update. I'll also add a log for the start().

File: src/lib/metrics.ts [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+   * Creates a new TimerMetricInstance
+   * @param metricTag used for logging to identify the metric
+   */
+  public constructor(metricTag: string) {
+    super();
+    this.metricTag = metricTag;
+  }
+
+  public getValue(): MetricValue {
+    if (this.startTimeMs != 0 && this.endTimeMs != 0) {
```
Comment (JackuB) [inline src/lib/metrics.ts:27]: Any reason to not use triple equal here?

**Review by JackuB — APPROVED** [review]

File: src/lib/metrics.ts (hunk 8 above)
Comment (maxjeffos) [reply src/lib/metrics.ts]: no

**Pull request merged**